File: src/main/java/com/github/alexandrecarlton/idea/settings/starter/HomeExpandingPathDeserializer.java
package com.github.alexandrecarlton.idea.settings.starter;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.deser.std.StdScalarDeserializer;

import java.io.IOException;
import java.nio.file.Path;

/**
 * A custom {@link Path} deserialiser that:
 *  - resolves the path relative to the project checkout.
 *  - expands {@code ~} to the user's home directory.
 * Note that this requires that the path exist on the user's filesystem.
 */
public class HomeExpandingPathDeserializer extends StdScalarDeserializer<Path> {

  private final Path basePath;

  public HomeExpandingPathDeserializer(Path basePath) {
    this(basePath, null);
  }

  private HomeExpandingPathDeserializer(Path basePath, Class<?> vc) {
    super(vc);
    this.basePath = basePath;
  }

  @Override
  public Path deserialize(JsonParser jp, DeserializationContext ctxt) throws IOException {
    return basePath.resolve(jp.getValueAsString()
        .replace("~", System.getProperty("user.home")))
        .toRealPath();
  }
}

File: src/main/kotlin/com/github/alexandrecarlton/idea/settings/starter/HomeExpandingPathDeserializer.kt
package com.github.alexandrecarlton.idea.settings.starter

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.deser.std.StdScalarDeserializer
import java.nio.file.Path

/**
 * A custom [Path] deserialiser that:
 * - resolves the path relative to the project checkout.
 * - expands `~` to the user's home directory.
 * Note that this requires that the path exist on the user's filesystem.
 */
// TODO: This should deserialise a File instead, once everything is in Kotlin.
class HomeExpandingPathDeserializer(private val basePath: Path, vc: Class<*>?) : StdScalarDeserializer<Path>(vc) {

    constructor(basePath: Path) : this(basePath, null)

    override fun deserialize(jp: JsonParser, ctxt: DeserializationContext): Path =
        basePath.resolve(jp.valueAsString
            .replace("~", System.getProperty("user.home")))
            // TODO: Remove toRealPath; just call it when comparing modules.
            // Causing us too many issues otherwise.
            .toRealPath()
}
